File: src-tauri/src/codegen/render.rs
use std::collections::{HashMap, HashSet};

use anyhow::Context;
use minijinja::{Environment, Value};
use rust_embed::RustEmbed;
use serde_json::json;

use crate::model::types::{
    Control, EncoderMode, JoystickAxis, Project, SelectorOp, pin_to_arduino_num,
};
use crate::pins::profile::profile_for;

use super::button_index::{assign_button_indices, total_button_count};
use super::emitter::GeneratedBoard;

#[derive(RustEmbed)]
#[folder = "templates/"]
struct Templates;

pub fn render_board(project: &Project, board_id: &str) -> anyhow::Result<GeneratedBoard> {
    let board = project
        .boards
        .iter()
        .find(|b| b.id == board_id)
        .ok_or_else(|| anyhow::anyhow!("Board '{}' not found", board_id))?;

    let profile = profile_for(&board.board_type);
    let button_map = assign_button_indices(project, board_id);
    let total_buttons = total_button_count(project, board_id);

    let board_controls: Vec<&Control> = project
        .controls
        .iter()
        .filter(|c| c.board_id() == board_id)
        .collect();

    let used_axes = collect_used_axes(&board_controls);

    let ctx = build_context(project, board_id, &board_controls, &button_map, total_buttons, &used_axes, &profile)?;

    let env = build_env()?;

    let platformio_ini = env
        .get_template("platformio.ini.jinja2")
        .context("loading platformio.ini template")?
        .render(&ctx)
        .context("rendering platformio.ini")?;

    let main_cpp = env
        .get_template("main.cpp.jinja2")
        .context("loading main.cpp template")?
        .render(&ctx)
        .context("rendering main.cpp")?;

    let board_json = env
        .get_template("board.json.jinja2")
        .context("loading board.json template")?
        .render(&ctx)
        .context("rendering board.json")
        .ok();

    Ok(GeneratedBoard {
        platformio_ini,
        main_cpp,
        board_json,
    })
}

fn build_env() -> anyhow::Result<Environment<'static>> {
    let mut env = Environment::new();
    env.set_loader(|name| match Templates::get(name) {
        Some(f) => Ok(Some(
            std::str::from_utf8(f.data.as_ref())
                .map(|s| s.to_string())
                .map_err(|e| minijinja::Error::new(
                    minijinja::ErrorKind::InvalidOperation,
                    e.to_string(),
                ))?,
        )),
        None => Ok(None),
    });
    Ok(env)
}

fn collect_used_axes(controls: &[&Control]) -> HashSet<String> {
    let mut axes = HashSet::new();
    for control in controls {
        match control {
            Control::Analog(a) => {
                axes.insert(axis_to_string(&a.analog.axis));
            }
            Control::Encoder(e) if e.encoder.mode == EncoderMode::Axis => {
                if let Some(axis) = &e.encoder.axis {
                    axes.insert(axis_to_string(axis));
                }
            }
            _ => {}
        }
    }
    axes
}

fn axis_to_string(axis: &JoystickAxis) -> String {
    match axis {
        JoystickAxis::X => "X",
        JoystickAxis::Y => "Y",
        JoystickAxis::Z => "Z",
        JoystickAxis::Rx => "Rx",
        JoystickAxis::Ry => "Ry",
        JoystickAxis::Rz => "Rz",
        JoystickAxis::Slider1 => "Slider1",
        JoystickAxis::Slider2 => "Slider2",
    }
    .to_string()
}

fn build_context(
    project: &Project,
    board_id: &str,
    controls: &[&Control],
    button_map: &HashMap<(String, usize), u32>,
    total_buttons: u32,
    used_axes: &HashSet<String>,
    profile: &crate::pins::profile::BoardProfile,
) -> anyhow::Result<Value> {
    let board = project.boards.iter().find(|b| b.id == board_id).unwrap();

    let mut button_controls = Vec::new();
    let mut switch_controls = Vec::new();
    let mut selector_controls = Vec::new();
    let mut encoder_controls = Vec::new();
    let mut interrupt_encoder_ids: HashSet<String> = HashSet::new();
    let mut polling_encoder_ids: HashSet<String> = HashSet::new();
    let mut analog_controls = Vec::new();

    for control in controls {
        let base = control.base();
        let cid = base.id.clone();

        match control {
            Control::Button(b) => {
                button_controls.push(json!({
                    "id": b.base.id,
                    "label": b.base.label,
                    "pin_num": pin_to_arduino_num(&b.pin.pin),
                    "inverted": b.pin.inverted,
                    "button_start": button_map.get(&(cid.clone(), 0)).copied().unwrap_or(0),
                }));
            }
            Control::Switch(s) => {
                switch_controls.push(json!({
                    "id": s.base.id,
                    "label": s.base.label,
                    "pin_num": pin_to_arduino_num(&s.pin.pin),
                    "inverted": s.pin.inverted,
                    "on_label": s.on_label,
                    "off_label": s.off_label,
                    "button_on": button_map.get(&(cid.clone(), 0)).copied().unwrap_or(0),
                    "button_off": button_map.get(&(cid.clone(), 1)).copied().unwrap_or(1),
                }));
            }
            Control::Selector(s) => {
                let positions: Vec<_> = s
                    .positions
                    .iter()
                    .enumerate()
                    .map(|(i, pos)| {
                        let pins: Vec<_> = pos
                            .pins
                            .iter()
                            .map(|pr| json!({
                                "pin_num": pin_to_arduino_num(&pr.pin),
                                "inverted": pr.inverted,
                            }))
                            .collect();
                        let op = match pos.op {
                            None => "single",
                            Some(SelectorOp::And) => "and",
                            Some(SelectorOp::Or) => "or",
                        };
                        json!({
                            "label": pos.label,
                            "pins": pins,
                            "op": op,
                            "button_idx": button_map.get(&(cid.clone(), i)).copied().unwrap_or(0),
                        })
                    })
                    .collect();

                selector_controls.push(json!({
                    "id": s.base.id,
                    "label": s.base.label,
                    "positions": positions,
                }));
            }
            Control::Encoder(e) => {
                let enc = &e.encoder;
                let use_interrupt = profile.interrupt_pins.contains(&enc.pin_a)
                    && profile.interrupt_pins.contains(&enc.pin_b);

                if use_interrupt {
                    interrupt_encoder_ids.insert(cid.clone());
                } else {
                    polling_encoder_ids.insert(cid.clone());
                }

                let mut ec = json!({
                    "id": e.base.id,
                    "label": e.base.label,
                    "pin_a_num": pin_to_arduino_num(&enc.pin_a),
                    "pin_b_num": pin_to_arduino_num(&enc.pin_b),
                    "pin_a": enc.pin_a,
                    "pin_b": enc.pin_b,
                    "counts_per_detent": enc.counts_per_detent,
                    "mode": match enc.mode { EncoderMode::Buttons => "buttons", EncoderMode::Axis => "axis" },
                    "use_interrupt": use_interrupt,
                });

                if enc.mode == EncoderMode::Buttons {
                    ec["presses_per_detent"] = json!(enc.presses_per_detent.unwrap_or(1));
                    ec["pulse_ms"] = json!(enc.pulse_ms.unwrap_or(20));
                    ec["button_cw"] = json!(button_map.get(&(cid.clone(), 0)).copied().unwrap_or(0));
                    ec["button_ccw"] = json!(button_map.get(&(cid.clone(), 1)).copied().unwrap_or(1));
                } else if let Some(axis) = &enc.axis {
                    ec["axis_setter"] = json!(axis.setter_method());
                    ec["delta_per_step"] = json!(enc.delta_per_step.unwrap_or(1));
                    ec["axis_min"] = json!(enc.min.unwrap_or(0));
                    ec["axis_max"] = json!(enc.max.unwrap_or(1023));
                    ec["wrap"] = json!(enc.wrap.unwrap_or(false));
                }

                encoder_controls.push(ec);
            }
            Control::Analog(a) => {
                let ana = &a.analog;
                analog_controls.push(json!({
                    "id": a.base.id,
                    "label": a.base.label,
                    "pin_num": pin_to_arduino_num(&ana.pin),
                    "axis_setter": ana.axis.setter_method(),
                    "in_min": ana.in_min,
                    "in_max": ana.in_max,
                    "out_min": ana.out_min,
                    "out_max": ana.out_max,
                    "invert": ana.invert,
                    "deadzone": ana.deadzone,
                    "smoothing": ana.smoothing,
                }));
            }
        }
    }

    let interrupt_encoders: Vec<_> = encoder_controls
        .iter()
        .filter(|e| e["use_interrupt"].as_bool().unwrap_or(false))
        .cloned()
        .collect();

    let polling_encoders: Vec<_> = encoder_controls
        .iter()
        .filter(|e| !e["use_interrupt"].as_bool().unwrap_or(false))
        .cloned()
        .collect();

    let all_axes = ["X", "Y", "Z", "Rx", "Ry", "Rz", "Slider1", "Slider2"];
    let axis_flags: serde_json::Map<String, serde_json::Value> = all_axes
        .iter()
        .map(|a| (a.to_string(), json!(used_axes.contains(*a))))
        .collect();

    let env_name = board.id.replace('-', "_");
    let board_type_str = match board.board_type {
        crate::model::types::BoardType::Leonardo => "leonardo",
        crate::model::types::BoardType::Micro => "micro",
        crate::model::types::BoardType::ProMicro => "pro_micro",
    };

    Ok(Value::from_serialize(json!({
        "board": {
            "id": board.id,
            "env_name": env_name,
            "name": board.name,
            "board_type": board_type_str,
            "identity": {
                "usbProduct": board.identity.usb_product,
                "usbVid": board.identity.usb_vid,
                "usbPid": board.identity.usb_pid,
                "usbVidHex": format!("0x{:04X}", board.identity.usb_vid),
                "usbPidHex": format!("0x{:04X}", board.identity.usb_pid),
            }
        },
        "total_buttons": total_buttons,
        "axis_flags": axis_flags,
        "has_interrupt_encoders": !interrupt_encoders.is_empty(),
        "button_controls": button_controls,
        "switch_controls": switch_controls,
        "selector_controls": selector_controls,
        "encoder_controls": encoder_controls,
        "interrupt_encoders": interrupt_encoders,
        "polling_encoders": polling_encoders,
        "analog_controls": analog_controls,
        "has_encoders": !encoder_controls.is_empty(),
        "has_analog": !analog_controls.is_empty(),
    })))
}

File: electron/engine/render.ts
// Firmware rendering — ported from server/src/codegen/render.rs.
// Builds the same template context shape the (now nunjucks) templates expect,
// then renders platformio.ini / main.cpp / board.json.

import nunjucks from "nunjucks";

import type { GeneratedBoard, JoystickAxis, Project } from "./types";
import { pinToArduinoNum } from "./model";
import { profileFor } from "./pins";
import { assignButtonIndices, totalButtonCount } from "./buttonIndex";
import { MAIN_CPP, PLATFORMIO_INI, BOARD_JSON } from "./templates";

const env = new nunjucks.Environment(null, {
  autoescape: false, // generating C++/INI/JSON, not HTML
  trimBlocks: true,
  lstripBlocks: true,
  throwOnUndefined: false,
});

const ALL_AXES: JoystickAxis[] = ["X", "Y", "Z", "Rx", "Ry", "Rz", "Slider1", "Slider2"];

/** Joystick library setter method for an axis (ported from JoystickAxis::setter_method). */
function axisSetter(axis: JoystickAxis): string {
  switch (axis) {
    case "X":
      return "setXAxis";
    case "Y":
      return "setYAxis";
    case "Z":
      return "setZAxis";
    case "Rx":
      return "setRxAxis";
    case "Ry":
      return "setRyAxis";
    case "Rz":
      return "setRzAxis";
    case "Slider1":
      return "setThrottle";
    case "Slider2":
      return "setRudder";
  }
}

function hex4(n: number): string {
  return `0x${n.toString(16).toUpperCase().padStart(4, "0")}`;
}

export function renderBoard(project: Project, boardId: string): GeneratedBoard {
  const board = project.boards.find((b) => b.id === boardId);
  if (!board) {
    throw new Error(`Board '${boardId}' not found`);
  }

  const profile = profileFor(board.type);
  const buttonMap = assignButtonIndices(project, boardId);
  const totalButtons = totalButtonCount(project, boardId);
  const boardControls = project.controls.filter((c) => c.boardId === boardId);

  const btn = (id: string, sub: number): number => buttonMap.get(id)?.get(sub) ?? sub;

  const buttonControls: unknown[] = [];
  const switchControls: unknown[] = [];
  const selectorControls: unknown[] = [];
  const encoderControls: Record<string, unknown>[] = [];
  const analogControls: unknown[] = [];
  const usedAxes = new Set<string>();

  for (const control of boardControls) {
    const cid = control.id;
    switch (control.kind) {
      case "button":
        if (!control.pin) break;
        buttonControls.push({
          id: control.id,
          label: control.label,
          pin_num: pinToArduinoNum(control.pin.pin),
          inverted: control.pin.inverted,
          button_start: btn(cid, 0),
        });
        break;
      case "switch":
        if (!control.pin) break;
        switchControls.push({
          id: control.id,
          label: control.label,
          pin_num: pinToArduinoNum(control.pin.pin),
          inverted: control.pin.inverted,
          on_label: control.onLabel,
          off_label: control.offLabel,
          button_on: btn(cid, 0),
          button_off: buttonMap.get(cid)?.get(1) ?? 1,
        });
        break;
      case "selector": {
        const positions = control.positions.map((pos, i) => ({
          label: pos.label,
          pins: pos.pins.map((pr) => ({
            pin_num: pinToArduinoNum(pr.pin),
            inverted: pr.inverted,
          })),
          op: pos.op === null || pos.op === undefined ? "single" : pos.op,
          button_idx: btn(cid, i),
        }));
        selectorControls.push({ id: control.id, label: control.label, positions });
        break;
      }
      case "encoder": {
        if (!control.encoder) break;
        const enc = control.encoder;
        const useInterrupt =
          profile.interruptPins.includes(enc.pinA) &&
          profile.interruptPins.includes(enc.pinB);
        const ec: Record<string, unknown> = {
          id: control.id,
          label: control.label,
          pin_a_num: pinToArduinoNum(enc.pinA),
          pin_b_num: pinToArduinoNum(enc.pinB),
          pin_a: enc.pinA,
          pin_b: enc.pinB,
          counts_per_detent: enc.countsPerDetent,
          mode: enc.mode,
          use_interrupt: useInterrupt,
        };
        if (enc.mode === "buttons") {
          ec.presses_per_detent = enc.pressesPerDetent ?? 1;
          ec.pulse_ms = enc.pulseMs ?? 20;
          ec.button_cw = btn(cid, 0);
          ec.button_ccw = buttonMap.get(cid)?.get(1) ?? 1;
        } else if (enc.axis != null) {
          usedAxes.add(enc.axis);
          ec.axis_setter = axisSetter(enc.axis);
          ec.delta_per_step = enc.deltaPerStep ?? 1;
          ec.axis_min = enc.min ?? 0;
          ec.axis_max = enc.max ?? 1023;
          ec.wrap = enc.wrap ?? false;
        }
        encoderControls.push(ec);
        break;
      }
      case "analog": {
        if (!control.analog) break;
        const ana = control.analog;
        usedAxes.add(ana.axis);
        analogControls.push({
          id: control.id,
          label: control.label,
          pin_num: pinToArduinoNum(ana.pin),
          axis_setter: axisSetter(ana.axis),
          in_min: ana.inMin,
          in_max: ana.inMax,
          out_min: ana.outMin,
          out_max: ana.outMax,
          invert: ana.invert,
          deadzone: ana.deadzone ?? null,
          smoothing: ana.smoothing ?? null,
        });
        break;
      }
    }
  }

  const interruptEncoders = encoderControls.filter((e) => e.use_interrupt === true);
  const pollingEncoders = encoderControls.filter((e) => e.use_interrupt !== true);

  const axisFlags = Object.fromEntries(ALL_AXES.map((a) => [a, usedAxes.has(a)]));

  const envName = board.id.replace(/-/g, "_");
  const buildVariant =
    board.type === "pro_micro" ? "sparkfun_promicro" : board.type;

  const ctx = {
    board: {
      id: board.id,
      env_name: envName,
      name: board.name,
      board_type: board.type,
      build_variant: buildVariant,
      identity: {
        usbProduct: board.identity.usbProduct,
        usbVid: board.identity.usbVid,
        usbPid: board.identity.usbPid,
        usbVidHex: hex4(board.identity.usbVid),
        usbPidHex: hex4(board.identity.usbPid),
      },
    },
    total_buttons: totalButtons,
    axis_flags: axisFlags,
    has_interrupt_encoders: interruptEncoders.length > 0,
    button_controls: buttonControls,
    switch_controls: switchControls,
    selector_controls: selectorControls,
    encoder_controls: encoderControls,
    interrupt_encoders: interruptEncoders,
    polling_encoders: pollingEncoders,
    analog_controls: analogControls,
    has_encoders: encoderControls.length > 0,
    has_analog: analogControls.length > 0,
  };

  const platformioIni = env.renderString(PLATFORMIO_INI, ctx);
  const mainCpp = env.renderString(MAIN_CPP, ctx);
  let boardJson: string | null = null;
  try {
    boardJson = env.renderString(BOARD_JSON, ctx);
  } catch {
    boardJson = null;
  }

  return { platformioIni, mainCpp, boardJson };
}
